### PR TITLE
solution_generator

### PR DESCRIPTION
**Pull request by arvidn**
This PR can be reviewed one commit at a time.

This patch switches the block generator function over to the rust implementation, which also supports serializing the program with backrefs (which is used after the hard fork activation). After the hard fork, this serialization replaces the existing block compression logic, which kicks in for blocks containing only vanilla transactions.

It also updates `BlockTools` to generate such blocks in tests, as well as fixes up tests that currently depend on the existing compression logic.

### CODE REVIEW

**Comment by arvidn** [comment]
```
Diff Coverage
Diff: origin/main...HEAD, staged and unstaged changes
-------------
chia/full_node/bundle_tools.py (100%)
chia/full_node/full_node_api.py (14.3%): Missing lines 777-778,780-781,785,789
chia/simulator/block_tools.py (100%)
chia/types/blockchain_format/program.py (100%)
tests/blockchain/test_blockchain.py (100%)
tests/generator/test_compression.py (100%)
```

File: chia/full_node/full_node_api.py [hunk 1]
```diff
@@ -767,16 +771,22 @@ async def declare_proof_of_space(
...
                         aggregate_signature = spend_bundle.aggregated_signature
-                        if self.full_node.full_node_store.previous_generator is not None:
-                            self.log.info(
-                                f"Using previous generator for height "
-                                f"{self.full_node.full_node_store.previous_generator}"
-                            )
-                            block_generator = best_solution_generator_from_template(
-                                self.full_node.full_node_store.previous_generator, spend_bundle
-                            )
+                        # when the hard fork activates, block generators are
```
Comment (arvidn) [inline chia/full_node/full_node_api.py:774]: it appears we don't have any test for farming blocks, which is a bit disconcerting

**Comment by arvidn** [comment]
I think it *is* a problem that we don't have a test that exercises farming a block. However, I don't think this PR can be held up for that. The section of code introduced for farming a block is the same as the blocks added to `BlockTools` (which is how we farm blocks in simulators), and that's tested.

If people feel worried about it, I'm fine with removing the farming code, since it won't take effect until the hard fork anyway, we'll have more opportunities to update this later.

**Review by wjblanke — APPROVED** [review]
aok

**Review by xdustinface — APPROVED** [review]
Looks good, left few nit comments, feel free to close them. And, i probably miss something here but i still wonder, for the other places where we `run_chia_program` we don't want to pass the `ALLOW_BACKREFS`?

File: tests/blockchain/test_blockchain.py [hunk 2]
```diff
@@ -2506,7 +2511,14 @@ async def test_invalid_transactions_ref_list(self, empty_blockchain, bt):
             previous_generator=generator_arg,
         )
         block = blocks[-1]
-        assert len(block.transactions_generator_ref_list) > 0
+        if consensus_mode == Mode.HARD_FORK_2_0:
+            # once the hard for activates, we don't use this form of block
+            # compression anymore
+            assert len(block.transactions_generator_ref_list) == 0
+            # the remaining tests assume a reflist
+            return
```
[the remark this answers is not in the document]
Comment (arvidn) [reply tests/blockchain/test_blockchain.py:2519]: yeah, I tried to keep this patch small, I'm guessing we'd want to split these tests up into two probably, in the future

File: chia/simulator/block_tools.py [hunk 3]
```diff
@@ -720,14 +721,20 @@ def get_consecutive_blocks(
...
-                                    previous_generator, transaction_data
-                                )
+                            if start_height >= constants.HARD_FORK_HEIGHT:
+                                block_generator = simple_solution_generator_backrefs(transaction_data)
+                                previous_generator = None
                             else:
-                                block_generator = simple_solution_generator(transaction_data)
-                                if type(previous_generator) is list:
-                                    block_generator = BlockGenerator(block_generator.program, [], previous_generator)
+                                if type(previous_generator) is CompressorArg:
```
[the remark this answers is not in the document]
Comment (arvidn) [reply chia/simulator/block_tools.py:728]: yeah, that's true. that would save indentation and diff. I was just thinking of the hard-fork case followed by the previous logic verbatim

File: tests/generator/test_compression.py [hunk 4]
```diff
@@ -167,12 +168,13 @@ def test_get_removals_for_single_coin(self) -> None:
         assert bytes(spend_info.puzzle) == bytes(sb.coin_spends[0].puzzle_reveal)
         assert bytes(spend_info.solution) == bytes(sb.coin_spends[0].solution)
 
-    def test_spend_byndle_coin_spend(self) -> None:
-        for i in range(0, 10):
-            sb: SpendBundle = make_spend_bundle(i)
-            cs1 = SExp.to(spend_bundle_to_coin_spend_entry_list(sb)).as_bin()  # pylint: disable=E1101
-            cs2 = spend_bundle_to_serialized_coin_spend_entry_list(sb)
-            assert cs1 == cs2
+        # test with backrefs (2.0 hard-fork)
```
[the remark this answers is not in the document]
Comment (arvidn) [reply tests/generator/test_compression.py:171]: the current compression logic is quite simplistic and works in a fundamentally different way compared to the new (backrefs) one. The backrefs (post hard fork) compression is self-contained whereas the current one finds a previous block in the chain to refer back to

**Pull request merged**